OpenRobertaServer/src/main/java/de/fhg/iais/roberta/util/archiver/ProgramsXmlArchiver.java

```diff
@@ -48,20 +49,22 @@ public void putProgram(Program program, String configuration) {
     public ByteArrayOutputStream getArchive() throws Exception {
         openStreams();
         for ( Pair<Program, String> progConf : programConfigList ) {
-            Program program = progConf.getFirst();
-            String config = progConf.getSecond();
-            String programXml = buildProgramXml(program, config);
-            String path = getProgramPath(program, config);
-
-            //add program.xml to Zip
-            zos.putNextEntry(new ZipEntry(path));
-            zos.write(programXml.getBytes());
-            zos.closeEntry();
+            addProgramToZip(progConf.getFirst(), progConf.getSecond());
         }
         closeStreams();
```

Comment:
closeStreams is not called if addProgramToZip throws an exception. A try-finally is needed here. With a bit of tweaking, it may be possible, to use a try-resource for your 2 streams.